### PR TITLE
tools: sof_perf_analyzer: fix widget ID parsing error

### PR DESCRIPTION
**Pull request by btian1**
In existing code, we parse component name from the line
that contains 'Create widget', and component ID from the
next line, which normally is the ipc tx message:

Create widget host-copier.0.playback instance 0 - pipe 1 - core 0
ipc tx      : 0x40000004|0x15: MOD_INIT_INSTANCE [data size: 84]

However, due to the concurrent nature of linux kernel, the
assumption that `Create widget` message and `ipc tx` message
are always adjacent is wrong. An exception will be raised
if the next line is not the 'ipc tx' message.

The first step to fix the issue is to make a kernel PR to print
all the information we need in a single 'Create widget' line,
thus to avoid the requirement for multiple lines of kernel message, 
see https://github.com/thesofproject/linux/pull/4709.

And next，this patch will parse the new 'Create widget' line to
get component name and ID.

### CODE REVIEW

File: tools/sof_perf_analyzer.py [hunk 1]
```diff
@@ -204,8 +205,29 @@ def process_kmsg_file():
...
+                    print("no valid widget ID found, try with bind information")
+                    not_found[widget_name] = ppln_id
+                    continue
                 comp_name[str(Component(ppln_id, widget_id))] = widget_name
+            elif match_obj := re.search(r' bind ', line):
+                if not_found:
+                    # parse current line to get widget name
+                    span_end_pos = match_obj.span()[1] - 1
+                    line_split = line[span_end_pos + 1:].split()
+                    widget_name = line_split[0].split(':')[0]
```
Comment (aiChaoSONG) [inline tools/sof_perf_analyzer.py:220]: we got two widget_name (s) and two widget_id (s) from the bind message, you need to handle both, what if the not_found widget is the second widget in bind message?

File: tools/sof_perf_analyzer.py [hunk 2]
```diff
@@ -204,8 +205,29 @@ def process_kmsg_file():
...
+                    # parse next line get widget ID
+                    next_line = next(f)
+                    widget_id = next_line.split('|')[0].split(':')[-1].strip()
+                    widget_id = int('0x' + widget_id[-6:], 16)
+                    # get ppln_id from dictionary
+                    if not_found[widget_name]:
+                        ppln_id = not_found[widget_name]
+                        comp_name[str(Component(ppln_id, widget_id))] = widget_name
+                        # delete not_fund
```
Comment (aiChaoSONG) [inline tools/sof_perf_analyzer.py:229]: comment is redundant, `del ...` is self-explanatory.

File: tools/sof_perf_analyzer.py [hunk 3]
```diff
@@ -204,8 +205,29 @@ def process_kmsg_file():
...
+                    not_found[widget_name] = ppln_id
+                    continue
                 comp_name[str(Component(ppln_id, widget_id))] = widget_name
+            elif match_obj := re.search(r' bind ', line):
+                if not_found:
+                    # parse current line to get widget name
+                    span_end_pos = match_obj.span()[1] - 1
+                    line_split = line[span_end_pos + 1:].split()
+                    widget_name = line_split[0].split(':')[0]
+                    # parse next line get widget ID
```
Comment (aiChaoSONG) [inline tools/sof_perf_analyzer.py:221]: parse next line to get widget ID.

File: tools/sof_perf_analyzer.py [hunk 4]
```diff
@@ -204,8 +205,29 @@ def process_kmsg_file():
...
+                    next_line = next(f)
+                    widget_id = next_line.split('|')[0].split(':')[-1].strip()
+                    widget_id = int('0x' + widget_id[-6:], 16)
+                    # get ppln_id from dictionary
+                    if not_found[widget_name]:
+                        ppln_id = not_found[widget_name]
```
Comment (aiChaoSONG) [inline tools/sof_perf_analyzer.py:227]: `if ppln_id := not_found[widget_name]`, this removes the repetition of `not_found[widget_name]`

**Review by aiChaoSONG — CHANGES_REQUESTED** [review]

File: tools/sof_perf_analyzer.py [hunk 5]
```diff
@@ -204,8 +205,29 @@ def process_kmsg_file():
...
                 # convert to the same ID format in mtrace
-                widget_id = int('0x' + widget_id[-6:], 16)
+                try:
+                    widget_id = int('0x' + widget_id[-6:], 16)
+                except ValueError:
+                    print("no valid widget ID found, try with bind information")
+                    not_found[widget_name] = ppln_id
+                    continue
                 comp_name[str(Component(ppln_id, widget_id))] = widget_name
+            elif match_obj := re.search(r' bind ', line):
```
[the remark this answers is not in the document]
Comment (btian1) [reply tools/sof_perf_analyzer.py:215]: it is not search 'bind', it is ' bind ', with current dmesg logs, all the search is valid
this is simple and works, how about keep as for now? since this works, it can keep implementation
simple.
image

File: tools/sof_perf_analyzer.py (hunk 1 above)
Comment (btian1) [reply tools/sof_perf_analyzer.py:220]: the code are making sure we are getting the first widget name and first widget ID.
as you see, there is [0] for widget_name, we always use first to match the case.

File: tools/sof_perf_analyzer.py (hunk 4 above)
Comment (btian1) [reply tools/sof_perf_analyzer.py:227]: at this moment, ppln_id is still valid? 
if not_found[widget_name]
this condition means: with current parsed widget_name, if meet this condition, means we need setup dictionary for this widget, any logic wrong?

File: tools/sof_perf_analyzer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (marc-hb) [reply tools/sof_perf_analyzer.py:222]: > with current dmesg logs, all the search is valid

This is called "programming by coincidence"

File: tools/sof_perf_analyzer.py [hunk 6]
```diff
@@ -182,11 +182,19 @@ def process_kmsg_file():
     [  334.818442] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3:
     ipc tx      : 0x40000004|0x15: MOD_INIT_INSTANCE [data size: 84]
 
+    However, for some cases, the above lines does not contain widget ID information, in this
```
Comment (marc-hb) [inline tools/sof_perf_analyzer.py:185]: Which cases? And what exactly is missing in these cases and why? An entire line and if yes which one? A part of a line and if yes which part? You must be much more specific, otherwise it's impossible to review the code without knowing what it is trying to parse.

File: tools/sof_perf_analyzer.py [hunk 7]
```diff
@@ -204,8 +212,31 @@ def process_kmsg_file():
                 next_line = next(f)
                 widget_id = next_line.split('|')[0].split(':')[-1].strip()
                 # convert to the same ID format in mtrace
-                widget_id = int('0x' + widget_id[-6:], 16)
+                try:
+                    widget_id = int('0x' + widget_id[-6:], 16)
+                except ValueError:
+                    print(f'no valid widget ID found for {widget_name}, try with bind message')
+                    not_found[widget_name] = ppln_id
+                    continue
```
Comment (marc-hb) [inline tools/sof_perf_analyzer.py:220]: Don't mix `next()` and `continue` in the same `for` loop, this is incredibly hard to follow.

**Review by marc-hb — CHANGES_REQUESTED** [review]

**Comment by marc-hb** [comment]
Long before we search the info in a different place, could we have a "quicker fix" where the entire script does not crash but replaces the ID with some "Unknown" category or something?

File: tools/sof_perf_analyzer.py (hunk 7 above)
Comment (marc-hb) [reply tools/sof_perf_analyzer.py:220]: BTW there `next_line_if_there_is_one = next(f, None)` if needed. It should probably be used here like this:

```python
next_line = next(f, None)
if not next_line:
   # use some "NotFound" placeholder
   # so the code does not crash later
   # other cleanup if needed.
   raise StopIteration

```

File: tools/sof_perf_analyzer.py (hunk 6 above)
Comment (btian1) [reply tools/sof_perf_analyzer.py:185]: https://github.com/thesofproject/sof-test/issues/1108

File: tools/sof_perf_analyzer.py (hunk 7 above)
Comment (btian1) [reply tools/sof_perf_analyzer.py:223]: 1. continue is correct, still need check the following lines in for loop.
2. ok

File: tools/sof_perf_analyzer.py (hunk 6 above)
Comment (marc-hb) [reply tools/sof_perf_analyzer.py:182]: Thanks, I had heard about 1108 but besides for extraordinary complex situations anyone with basic Python knowledge should be able to understand the code without opening bugs which are not even linked from the source. This code is not an extraordinary complex situation, it's just parsing some logs so it must be easy to follow. Thanks for adding an example!

```suggestion
However, for some cases, the line after "Creating" does not contain widget ID information, in this
```

Also explain why the line after "Creating" does not always contain widget ID information, let's not "program by coincidence".

File: tools/sof_perf_analyzer.py (hunk 7 above)
Comment (marc-hb) [reply tools/sof_perf_analyzer.py:220]: > 1. continue is correct, still need check the following lines in for loop.

I didn't say it was incorrect. I said it was unreadable.

Now I understand the `bind` fallback is actually in a totally different part of the logs. In that case use two different `for` loops. There's no need to look for both at the same time, it makes the code incredibly hard to follow.

**Review by marc-hb — CHANGES_REQUESTED** [review]

File: tools/sof_perf_analyzer.py [hunk 8]
```diff
@@ -274,8 +271,8 @@ def analyze_perf_info():
     perf_stats.columns = ['CPU_AVG(MIN)', 'CPU_AVG(AVG)', 'CPU_AVG(MAX)',
                           'CPU_PEAK(MIN)', 'CPU_PEAK(AVG)', 'CPU_PEAK(MAX)']
     perf_stats['PEAK(MAX)/AVG(AVG)'] = perf_stats['CPU_PEAK(MAX)'] / perf_stats['CPU_AVG(AVG)']
-    comp_peak_avg_cycles = perf_info.groupby('COMP_ID')['CPU_PEAK'].mean()
-    perf_stats['MODULE_CPC'] = comp_peak_avg_cycles * CPC_MARGIN
+    comp_peak_max_cycles = perf_info.groupby('COMP_ID')['CPU_PEAK'].max()
+    perf_stats['MODULE_CPC'] = comp_peak_max_cycles
```
Comment (aiChaoSONG) [inline tools/sof_perf_analyzer.py:275]: `perf_stats['MODULE_CPC'] = perf_info.groupby('COMP_ID')['CPU_PEAK'].max()` single line is good.

Let's make this commit a standalone PR, because
- it has nothing to do with widget id issue
- it can be merged quickly

**Comment by aiChaoSONG** [comment]
@btian1 For the widget id issue, I am afraid we have to refine code and comments according Marc's review comments. Let's sync tomorrow.

File: tools/sof_perf_analyzer.py (hunk 7 above)
Comment (aiChaoSONG) [reply tools/sof_perf_analyzer.py:220]: @btian1 Move `comp_name[str(Component(ppln_id, widget_id))] = widget_name` to the try block, and remove `continue`

**Review by aiChaoSONG — DISMISSED** [review]

File: tools/sof_perf_analyzer.py [hunk 9]
```diff
@@ -237,8 +271,7 @@ def analyze_perf_info():
     perf_stats.columns = ['CPU_AVG(MIN)', 'CPU_AVG(AVG)', 'CPU_AVG(MAX)',
                           'CPU_PEAK(MIN)', 'CPU_PEAK(AVG)', 'CPU_PEAK(MAX)']
     perf_stats['PEAK(MAX)/AVG(AVG)'] = perf_stats['CPU_PEAK(MAX)'] / perf_stats['CPU_AVG(AVG)']
-    comp_peak_avg_cycles = perf_info.groupby('COMP_ID')['CPU_PEAK'].mean()
-    perf_stats['MODULE_CPC'] = comp_peak_avg_cycles * CPC_MARGIN
+    perf_stats['MODULE_CPC'] = perf_info.groupby('COMP_ID')['CPU_PEAK'].max()
```
Comment (marc-hb) [inline tools/sof_perf_analyzer.py:274]: This second commit does not seem related at all to the first one. Is it?

File: tools/sof_perf_analyzer.py [hunk 10]
```diff
@@ -182,11 +179,19 @@ def process_kmsg_file():
     [  334.818442] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3:
     ipc tx      : 0x40000004|0x15: MOD_INIT_INSTANCE [data size: 84]
 
+    However, for some cases, the line after "Creating" does not contain widget ID information,
```
Comment (marc-hb) [inline tools/sof_perf_analyzer.py:182]: In which cases? Is that not a kernel bug that this PR would then hide?

cc: @plbossart , @ranj063

File: tools/sof_perf_analyzer.py [hunk 11]
```diff
@@ -201,11 +206,40 @@ def process_kmsg_file():
...
+                        # reach EOF, exit.
+                        break
+                    widget_ids = next_line.split(': ')[-2].split('|')
+                    widget_ids = int('0x' + widget_ids[0][-6:], 16), int(widget_ids[1], 16)
+
+                    widgets = dict(zip(widget_names, widget_ids))
+
+                    # search and set missed comp_name
+                    for widget_name, widget_id in widgets.items():
+                        if ppln_id := not_found.get(widget_name):
```
Comment (marc-hb) [inline tools/sof_perf_analyzer.py:240]: This `not_found` makes the code significantly more complicated for no obvious benefit, please get rid of it. It should be as simple as:

1) Use the `for line` loop to build some `id + name` dictionary. Don't touch `comp_name` yet. When you find a `bind` line, simply give `Creating` lines precedence in the temporary dictionary.

2) AFTER the `for line` loop, use the temporary and complete dictionary to create `comp_name`.

**Review by marc-hb — CHANGES_REQUESTED** [review]

File: tools/sof_perf_analyzer.py (hunk 11 above)
Comment (aiChaoSONG) [reply tools/sof_perf_analyzer.py:240]: @marc-hb Component class have pipeline_id and comp_id, two parts here. 
from `create widget` message we could parse widget(component) name and pipeline id, and from next line we parse comp_id. on next line parsing error, we only have pipeline id and widget name.

from bind message, we parse widget name, and from its next line, we parse comp_id. bind message is complementary information if we failed to parse some 'create message'. 

in both `create widget` and bind message parsing, two lines are involved. we cannot assume it is atomic. because the concurrent nature of the linux kernel.

We are use two types of information to get more Component class here.  there is still probability a component in the topology is missing in our comp_name dict, but the probability is low as we parse more and more information ('create widget', 'bind message', 'unbind message'), but in practice, with 'create widget' and bind message,  the error rate can be think as low as zero.

Comment (marc-hb) [reply tools/sof_perf_analyzer.py:240]: Thanks @aiChaoSONG for the explanation. The length of your explanation demonstrates that the data being parsed is complex for various reasons. It's not possible to review this parsing code for correctness without understanding the data it's trying to parse. So examples are very useful but they are not enough. I'm afraid this is one of these rare cases where it's very hard to make the code "self-explanatory" and where a lot more comments than usual are required in the source code. For sure a lot of what you just wrote should make its way to the source code.

People's assignments come and go and we cannot have only 1.5 people in the team who can understand this code (or any code for that matter)

"“Programs must be written for people to read, and only incidentally for machines to execute.” - this is especially true in Python code which is very slow to execute (but usually much faster to read).

Also, forcing oneself to explain the code (with comments) very often results in finding simplifications; writing comments is a good example of https://en.wikipedia.org/wiki/Rubber_duck_debugging

File: tools/sof_perf_analyzer.py (hunk 10 above)
Comment (plbossart) [reply tools/sof_perf_analyzer.py:182]: @btian1 what exactly is the problem statement? You are suggesting a solution without giving any context or alternate suggestion. That's not very helpful, I have no idea what this patch does.

**Review by plbossart — DISMISSED** [review]
no idea what this is about

**Comment by plbossart** [comment]
"When there is no widget ID in kernel logs, try to find the widget ID in bind related logs."

Has anyone tried to change the kernel logs to make the problem go away?

**Review by marc-hb — CHANGES_REQUESTED** [review]
> in both create widget and bind message parsing, two lines are involved. we cannot assume it is atomic. because the concurrent nature of the linux kernel.
> [...]
> there is still probability a component in the topology is missing in our comp_name dict, but the probability is low as we parse more and more information

I just had a quick chat with @plbossart and he told me that there has been zero effort to simplify the kernel logs... is this real?

File: tools/sof_perf_analyzer.py [hunk 12]
```diff
@@ -274,8 +271,7 @@ def analyze_perf_info():
     perf_stats.columns = ['CPU_AVG(MIN)', 'CPU_AVG(AVG)', 'CPU_AVG(MAX)',
                           'CPU_PEAK(MIN)', 'CPU_PEAK(AVG)', 'CPU_PEAK(MAX)']
     perf_stats['PEAK(MAX)/AVG(AVG)'] = perf_stats['CPU_PEAK(MAX)'] / perf_stats['CPU_AVG(AVG)']
-    comp_peak_avg_cycles = perf_info.groupby('COMP_ID')['CPU_PEAK'].mean()
-    perf_stats['MODULE_CPC'] = comp_peak_avg_cycles * CPC_MARGIN
+    perf_stats['MODULE_CPC'] = perf_info.groupby('COMP_ID')['CPU_PEAK'].max()
```
Comment (plbossart) [inline tools/sof_perf_analyzer.py:274]: it's also arguable that the CPC value needs to be e.g. 1.2 the max measured value to account for unknowns such as cache and IPC.

 CPC = measured_peak_cpu * CPC_MARGIN

File: tools/sof_perf_analyzer.py (hunk 10 above)
Comment (btian1) [reply tools/sof_perf_analyzer.py:182]: @plbossart , please check: https://github.com/thesofproject/sof-test/issues/1108
there is a possibility of log missed for kernel log output, in this case, this patch try to parse bind related info to get widget ID, widget name and so on.

@marc-hb , I checked kernel code, there should have no bug in kernel, just kernel log missing with a possibility:
https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc4.c#L222
@plbossart , please double confirm.

File: tools/sof_perf_analyzer.py (hunk 11 above)
Comment (btian1) [reply tools/sof_perf_analyzer.py:240]: @aiChaoSONG shall we go with itertools.pairwise？

File: tools/sof_perf_analyzer.py (hunk 12 above)
Comment (btian1) [reply tools/sof_perf_analyzer.py:274]: ok, peak is already much larger than avg(2-5 times), unless all module peak happen in the same time, then the budget should be ok.

File: tools/sof_perf_analyzer.py (hunk 9 above)
Comment (btian1) [reply tools/sof_perf_analyzer.py:274]: yes, if you want to have a separate PR, I will do it.

File: tools/sof_perf_analyzer.py (hunk 10 above)
Comment (plbossart) [reply tools/sof_perf_analyzer.py:185]: I will double-confirm that I don't understand the problem statement and why you are not trying to fix the issue (if it's real) at the source by updating the kernel logs.

Comment (btian1) [reply tools/sof_perf_analyzer.py:185]: Thanks, @plbossart , let me use logs to describe:
expected logs:
[61230.667725] kernel: snd_sof:sof_ipc4_widget_setup: sof-audio-pci-intel-mtl 0000:00:1f.3: Create widget mixin.20.1 instance 0 - pipe 20 - core 0
[61230.667728] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc tx      : 0x40000002|0xa: MOD_INIT_INSTANCE [data size: 40]

we can get name(mixin.20.1) and its module ID: 2(0x40000002).

unexpected logs(may not have errror in kernel):

[  262.012482] kernel: snd_sof:sof_ipc4_widget_setup: sof-audio-pci-intel-mtl 0000:00:1f.3: Create widget alh-copier.SDW0-Capture.0 instance 8 - pipe 11 - core 0
[  262.012486] kernel: snd_sof_intel_hda_common:hda_dsp_stream_trigger: sof-audio-pci-intel-mtl 0000:00:1f.3: FW Poll Status: reg[0xa0]=0x2024001e successful

[  261.979993] kernel: snd_sof:sof_ipc4_widget_setup: sof-audio-pci-intel-mtl 0000:00:1f.3: Create widget pipeline.31 instance 4 - pipe 31 - core 0
[  261.980114] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc tx reply: 0x33000000|0x1: GLB_SET_PIPELINE_STATE

[61230.948775] kernel: snd_sof:sof_ipc4_widget_setup: sof-audio-pci-intel-mtl 0000:00:1f.3: Create widget gain.15.1 instance 4 - pipe 15 - core 0
[61230.948777] kernel: snd_sof:sof_pcm_trigger: sof-audio-pci-intel-mtl 0000:00:1f.3: pcm: trigger stream 0 dir 0 cmd 1

so we get those info from bind related logs:
[61230.670335] kernel: snd_sof:sof_ipc4_route_setup: sof-audio-pci-intel-mtl 0000:00:1f.3: bind host-copier.2.playback:0 -> gain.20.1:0
[61230.670338] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc tx      : 0x45000004|0x6: MOD_BIND

get name(host-copier.2.playback), ID 4(0x45000004).

this is this patch's purpose.

Comment (plbossart) [reply tools/sof_perf_analyzer.py:185]: I am unable to follow what's missing or not right in the logs above.

It's rather disappointing that after multiple asks I still get this sort of raw dump of information without clarity on the problem statement. This is starting to be a waste of everyone's time if I am honest.

**Review by marc-hb — DISMISSED** [review]
> It's rather disappointing that after multiple asks I still get this sort of raw dump of information without clarity on the problem statement. This is starting to be a waste of everyone's time if I am honest. - @plbossart 


_About two weeks ago, 4 days after this PR was first submitted_, I asked this:
https://github.com/thesofproject/sof-test/pull/1122#discussion_r1382264896

> Which cases? And what exactly is missing in these cases and why? An entire line and if yes which one? A part of a line and if yes which part? You must be much more specific, otherwise it's impossible to review the code without knowing what it is trying to parse.

That question was marked as "resolved" without an answer.

At the same time I asked basically the same question in https://github.com/thesofproject/sof-test/issues/1108. No answer either.


It's _pointless_ to submit a fix to a problem that has not been analyzed, understood and explained. It's called "programming by coincidence".


1. **Why are the kernel logs incomplete, how and when?** 
1.  **Why can't the kernel logs be fixed instead?** Or at least made somewhat better.

Don't answer in this PR here, answer in the bug 1108

File: tools/sof_perf_analyzer.py (hunk 10 above)
Comment (bardliao) [reply tools/sof_perf_analyzer.py:185]: @plbossart I chatted with @btian1. The purpose is to connect the widget name and the IPC primary data. Like `mixin.20.1` <-> `0x40000002`.
The existing code get the information from dmesg log when it looks for `Create widget` and parse the log in 2 lines. It works in most cases, but as you know, the kernel is multi thread and there is no guarantee that the `ipc tx :` log will be right after the `Create widget` log. And the existing code will return error in that case.
The commit will use another key work `bind` to get the IPC primary data. @btian1 @aiChaoSONG Please correct me if I was wrong.

It should be easy to "solve" the issue if we can add msg->primary to 
```
dev_dbg(sdev->dev, "Create widget %s instance %d - pipe %d - core %d\n",
                swidget->widget->name, swidget->instance_id, swidget->pipeline_id, swidget->core);
```
Note that I don't mean that we need to fix it from kernel. But trying to explain the issue they are trying to fix.

Comment (marc-hb) [reply tools/sof_perf_analyzer.py:185]: > It should be easy to "solve" the issue if we can add msg->primary to

Sounds great to me! @plbossart ?

Comment (plbossart) [reply tools/sof_perf_analyzer.py:185]: The IPC is serialized so the order will not vary. There might be other things in between IPCs but they will be by construction based on different/irrelevant information.

So yes if we can change the logs there's no issue. This isn't an ABI, we can tune the logs to the needs of profiling tools.

Comment (ranj063) [reply tools/sof_perf_analyzer.py:185]: Sorry for jumping in a bit late. This is not an error in the kernel ogs nor an unexpected sequence of logs. When widgets in a pipeline are set up, the first widget in the pipeline sets up the pipeline before the widget itself gets set up. So please do not assume that the log ```dev_dbg(sdev->dev, "Create widget %s instance %d - pipe %d - core %d\n",...```. If you do , please fix it instead of assuming incorrect sequences.

Comment (plbossart) [reply tools/sof_perf_analyzer.py:185]: not able to follow your answer @ranj063, there's something missing in the sentence "do not assume that the log"

Comment (ujfalusi) [reply tools/sof_perf_analyzer.py:185]: The swidget is a Linux SOF concept and it represents all sorts of 'Sof WIDGETs'.
The pipeline is also a swidget and that is not  a module, it has no 'module ID'

Comment (ujfalusi) [reply tools/sof_perf_analyzer.py:185]: My proposal is that the kernel should print two version for swidget creation.
For 'modules': `Create widget for module: %s, module id: %u, instance id: %u, pipeline: %u, core: %u`
For 'pipelines': `Create widget for pipeline: %s, id: %u, instance id: %u, core: %u`

The pipeline_id is a Linux driver concept, it is good to have it out, but not sure what would be the best way.

File: tools/sof_perf_analyzer.py [hunk 13]
```diff
@@ -175,12 +175,14 @@ def process_trace_file():
...
-    [  334.818442] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3:
-    ipc tx      : 0x40000004|0x15: MOD_INIT_INSTANCE [data size: 84]
+    component name and ID are acquired from the line that contains 'Create widget',
+    and pipeline ID is acuquired from the line that contains 'Create pipeline'
+    Below are example logs for parsing:
+    [   59.622134] snd_sof:sof_ipc4_widget_setup: sof-audio-pci-intel-mtl 0000:00:1f.3:
+    Create pipeline pipeline.3 (pipe 3) - instance 3, core 0
+
+    [   59.622645] snd_sof:sof_ipc4_widget_setup: sof-audio-pci-intel-mtl 0000:00:1f.3:
+    Create widget host-copier.0.capture (pipe 3) - ID 4, instance 3, core 0
```
Comment (aiChaoSONG) [inline tools/sof_perf_analyzer.py:186]:
```
'''Process the dmesg to get component ID to component name mapping,
they are acquired from the line that contains 'Create widget':
    EXAMPLE_OF_CREATE_WIDGWT_LINE
By design in the kernel, pipeline ID is the instance ID of pipeline
widget, so it is acquired from the line that contains 'Create pipeline':
    EXAMPLE_OF_CREATE_WIDGET_LINE
'''
```

File: tools/sof_perf_analyzer.py [hunk 14]
```diff
@@ -190,21 +192,24 @@ def process_kmsg_file():
...
-                # firmware. Note that ppln_id variable will be assigned properly at pipeline widget
-                # creation, because it is always the first one to be created before all other
-                # widgets in the same pipeline and pipelines are created sequentially.
+                # with sof kernel PR: https://github.com/thesofproject/linux/pull/4709
+                # pipeline and module creation are using separate print in kernle log.
+                # So, for pipeline case, pipeline instance ID will be extracted from pipeline
+                # logs. For module creation, module instance ID and module ID will be
+                # extracted accordingly.
+                # ppln_id, widget_id and widget instance ID will be combined together to
+                # column COMP_ID as the unique ID for each module inside pipelines.
```
Comment (aiChaoSONG) [inline tools/sof_perf_analyzer.py:205]: keep my original comments, it explains why instance id of pipeline widget is the widget pipeline id for the widgets of the pineline

File: tools/sof_perf_analyzer.py [hunk 15]
```diff
@@ -190,21 +192,24 @@ def process_kmsg_file():
...
-                    ppln_id = line_split[2]
-                next_line = next(f)
-                widget_id = next_line.split('|')[0].split(':')[-1].strip()
-                # convert to the same ID format in mtrace
-                widget_id = int('0x' + widget_id[-6:], 16)
+                    ppln_id = int(line_split[5][:-1])
+                    continue
+
+                module_instance_id = int(line_split[7][:-1])
+                widget_id = int(line_split[5][:-1])
```
Comment (aiChaoSONG) [inline tools/sof_perf_analyzer.py:211]: same, explain `[:-1]` with a simple comment.

File: tools/sof_perf_analyzer.py [hunk 16]
```diff
@@ -190,21 +192,24 @@ def process_kmsg_file():
...
+                # extracted accordingly.
+                # ppln_id, widget_id and widget instance ID will be combined together to
+                # column COMP_ID as the unique ID for each module inside pipelines.
                 if widget_name.startswith('pipeline'):
-                    ppln_id = line_split[2]
-                next_line = next(f)
-                widget_id = next_line.split('|')[0].split(':')[-1].strip()
-                # convert to the same ID format in mtrace
-                widget_id = int('0x' + widget_id[-6:], 16)
+                    ppln_id = int(line_split[5][:-1])
```
Comment (aiChaoSONG) [inline tools/sof_perf_analyzer.py:207]: explain `[:-1]`, no one know why if you don't explain.
`remove ending comma(,) with [:-1]`

**Review by aiChaoSONG — CHANGES_REQUESTED** [review]
Commit title and body:
```
tools: sof_perf_analyzer: fix widget ID parsing error

In existing code, we parse component name from the line
that contains 'Create widget', and component ID from the
next line, which normally is the ipc tx message:

Create widget host-copier.0.playback instance 0 - pipe 1 - core 0
ipc tx      : 0x40000004|0x15: MOD_INIT_INSTANCE [data size: 84]

However, due to the concurrent nature of linux kernel, the
assumption that `Create widget` message and `ipc tx` message
are always adjacent is wrong. An exception will be raised
if the next line is not the 'ipc tx' message.

The first step to fix the issue is to make a kernel PR to print
all the information we need in a single 'Create widget' line,
thus to avoid the requirement for multiple lines of kernel message, 
see https://github.com/thesofproject/linux/pull/4709.

And next，this patch will parse the new 'Create widget' line to
get component name and ID.

Fixes: https://github.com/thesofproject/sof-test/issues/1108
```

Always describe the issue (what), root cause of the issue (why), and how it is fixed(how).

**Comment by aiChaoSONG** [comment]
see internal test report 35019， test result is good.

**Review by aiChaoSONG — APPROVED** [review]
